lib/plugin-api/hydrogen-provider.js

```diff
@@ -11,31 +11,38 @@
 * learning how to interact with Hydrogen in your own plugin.
 *
 * @class HydrogenProvider
-###
-module.exports =
-class HydrogenProvider
-    constructor: (@_hydrogen) ->
-        @_happy = true
+*/
+export default class HydrogenProvider {
+    constructor(_hydrogen) {
+        this._hydrogen = _hydrogen;
+        this._happy = true;
+    }
 
-    ###
+    /*
     * Calls your callback when the kernel has changed.
     * @param {Function} Callback
-    ###
-    onDidChangeKernel: (callback) ->
-        @_hydrogen.emitter.on 'did-change-kernel', (kernel) ->
-            if kernel?
-                callback kernel.getPluginWrapper()
-            else
-                callback null
+    */
+    onDidChangeKernel(callback) {
+        return this._hydrogen.emitter.on('did-change-kernel', function(kernel) {
+            if (kernel != null) {
+                return callback(kernel.getPluginWrapper());
+            } else {
+                return callback(null);
+            }
+        });
+    }
 
-    ###
+    /*
     * Get the `HydrogenKernel` of the currently active text editor.
     * @return {Class} `HydrogenKernel`
-    ###
-    getActiveKernel: ->
-        unless @_hydrogen.kernel?
-            grammar = @_hydrogen.editor.getGrammar()
-            language = @_hydrogen.kernelManager.getLanguageFor grammar
-            throw new Error "No running kernel for language `#{language}` found"
+    */
+    getActiveKernel() {
+        if (this._hydrogen.kernel == null) {
+            let grammar = this._hydrogen.editor.getGrammar();
+            let language = this._hydrogen.kernelManager.getLanguageFor(grammar);
+            throw new Error(`No running kernel for language \`${language}\` found`);
+        }
 
-        return @_hydrogen.kernel.getPluginWrapper()
+        return this._hydrogen.kernel.getPluginWrapper();
```

Comment:
> This is so that a plugin always gets the same wrapper object for the same kernel (same reference, not just equal-by-value).

Wouldn't you get the same wrapper if the wrapper is created by Kernel's constructor?

> Plugins need to be able to store metadata per-kernel.

That's a bad pattern, because API users don't have any guarantee to avoid name clashes.
Composition would be simple and safer pattern:
```js
let myH2Kernel = {
    kernel: kernelWrapper,
    myMetadata1: myMetadata1, 
    myMetadata2: myMetadata2, 
};